ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java

```diff
@@ -152,13 +154,18 @@ public NetworkClient(Selector selector, NetworkConfig networkConfig, NetworkMetr
         Port port = requestMetadata.requestInfo.getPort();
         String connId = connectionTracker.checkOutConnection(host, port);
         if (connId == null) {
-          if (connectionTracker.mayCreateNewConnection(host, port)) {
+          if (requestMetadata.pendingConnectionId == null && connectionTracker.mayCreateNewConnection(host, port)) {
             connId = selector.connect(new InetSocketAddress(host, port.getPort()), networkConfig.socketSendBufferBytes,
                 networkConfig.socketReceiveBufferBytes, port.getPortType());
             connectionTracker.startTrackingInitiatedConnection(host, port, connId);
-            logger.trace("Initiated a connection to {}:{} ", host, port);
+            requestMetadata.pendingConnectionId = connId;
+            pendingConnectionsToAssociatedRequests.put(connId, requestMetadata);
+            logger.trace("Initiated a connection to host {} port {} ", host, port);
           }
         } else {
+          if (requestMetadata.pendingConnectionId != null) {
+            requestMetadata.pendingConnectionId = null;
```

Comment:
+1 for gopal's comment. We might have a correctness issue
If this request initiated a connection but then received a connection that another request had initiated (which is possible), then if the connection it initiated fails, it will still get a disconnect notification when the connection it is actually using is not disconnected?